### PR TITLE
Support internvl-chat for pytorch engine

### PR DESCRIPTION
**Pull request by RunningLeon**
## Motivation

Support InternVL-Chat models for pytorch engine:
- [OpenGVLab/InternVL-Chat-V1-5](https://huggingface.co/OpenGVLab/InternVL-Chat-V1-5)
- [OpenGVLab/Mini-InternVL-Chat-2B-V1-5](https://huggingface.co/OpenGVLab/Mini-InternVL-Chat-2B-V1-5)
- [OpenGVLab/Mini-InternVL-Chat-4B-V1-5](https://huggingface.co/OpenGVLab/Mini-InternVL-Chat-4B-V1-5)

### Note
Required PRs
- #1641 
- #1825 

as requested in issue https://github.com/InternLM/lmdeploy/issues/1794
## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

None

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: lmdeploy/pytorch/engine/engine.py [hunk 1]
```diff
@@ -417,32 +418,30 @@ def __get_vlm_embeddings():
...
-                history_image_nums=history_image_nums,
-                history_image_token_lengths=history_image_token_lengths,
-                input_embeddings=input_embeddings,
-                input_embedding_indexing=input_embedding_indexing,
-                input_embedding_ranges=input_embedding_ranges)
+        # for inputs with embeddings
+        history_image_nums = None
+        history_image_token_lengths = None
+        # only for cogvlm
+        if self.model_config.model_arch == 'CogVLMForCausalLM':
```
Comment (grimoire) [inline lmdeploy/pytorch/engine/engine.py:425]: why remove task type check

File: lmdeploy/pytorch/engine/model_agent.py [hunk 2]
```diff
@@ -655,7 +655,7 @@ def _build_model(self,
             model_init_kwargs = dict(torch_dtype=torch_dtype,
                                      device_map=device)
             model_init_kwargs.update(self.model_config.init_kwargs)
-            if self.model_config.auto_model_cls is AutoModelForCausalLM:
+            if hasattr(self.model_config.auto_model_cls, 'from_config'):
```
Comment (grimoire) [inline lmdeploy/pytorch/engine/model_agent.py:658]: I would rather add `trust_remote_code` into `model_config.init_kwargs` so we do not need to check anything here.

**Comment by muhamad-rizki** [comment]
any news on this ?

File: lmdeploy/pytorch/engine/engine.py (hunk 1 above)
Comment (RunningLeon) [reply lmdeploy/pytorch/engine/engine.py:425]: because for llm models, it's also could pass embeddings when calling `engine.decode` method

Comment (grimoire) [reply lmdeploy/pytorch/engine/engine.py:425]: I would rather add a static condition to manually enable passing embedding. Build a dataclass is heavy even when all fields are None.

**Comment by RunningLeon** [comment]
@grimoire 
## performance on llama3-8b
cmd
```
python benchmark/profile_throughput.py ./ShareGPT_V3_unfiltered_cleaned_split.json /nvme/shared_data/llama3/Meta-Llama-3-8B-Instruct --backend pytorch -n 3000
```

### main branch
```
concurrency: 256
elapsed_time: 199.272s

first token latency(s)(min, max, ave): 1.238, 11.655, 2.881
per-token latency(s) percentile(50, 75, 95, 99): [0.04, 0.042, 0.278, 0.34]

number of prompt tokens: 676779
number of completion tokens: 612685
token throughput (completion token): 3074.616 token/s
token throughput (prompt + completion token): 6470.873 token/s
RPS (request per second): 15.055 req/s
RPM (request per minute): 903.288 req/min
--------------------------------------------------
```
### this PR
```
concurrency: 256
elapsed_time: 205.748s

first token latency(s)(min, max, ave): 1.033, 11.620, 2.934
per-token latency(s) percentile(50, 75, 95, 99): [0.041, 0.044, 0.279, 0.343]

number of prompt tokens: 676779
number of completion tokens: 612685
token throughput (completion token): 2977.838 token/s
token throughput (prompt + completion token): 6267.192 token/s
RPS (request per second): 14.581 req/s
RPM (request per minute): 874.855 req/min
--------------------------------------------------
```

**Review by grimoire — APPROVED** [review]

File: lmdeploy/model.py [hunk 3]
```diff
@@ -1458,6 +1458,45 @@ def match(cls, model_path: str) -> Optional[str]:
             return 'phi-3'
 
 
+@MODELS.register_module(name='internvl-phi3')
+class InternVLPhi3(BaseChatTemplate):
```
Comment (lvhan028) [inline lmdeploy/model.py:1462]: How about  deriving? from `Phi3Instruct`?

Comment (RunningLeon) [reply lmdeploy/model.py]: ok

**Comment by lvhan028** [comment]
```
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 60.00 MiB. GPU 5 has a total capacty of 79.14 GiB of which 5.19 MiB is free. Process 151190 has 78.15 GiB memory in use. Including non-PyTorch memory, this process has 992.00 MiB memory in use. Of the allocated memory 489.86 MiB is allocated by PyTorch, and 12.14 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

**Comment by lvhan028** [comment]
My test code is as follows, and the test command is `python examples/workspace/test_vl_pipeline.py /nvme/shared/InternVL-Chat-V1-5 --backend pytorch --cache_max_entry_count 0.5`
```python
from lmdeploy import pipeline, TurbomindEngineConfig, PytorchEngineConfig
from lmdeploy.vl import load_image
import fire
import time

def apply_turbomind(model_path, **kwargs):
    engine_config = TurbomindEngineConfig.from_dict({}, allow_none=True)

    for key, value in kwargs.items():
        if hasattr(TurbomindEngineConfig, key):
            setattr(engine_config, key, value)

    pipe = pipeline(model_path, backend_config=engine_config)

    return pipe


def apply_pytorch(model_path, **kwargs):
    engine_config = PytorchEngineConfig()
    for key, value in kwargs.items():
        print(key, value)
        if hasattr(PytorchEngineConfig, key):
            setattr(engine_config, key, value)
            
    pipe = pipeline(model_path, backend_config=engine_config)
    
    return pipe


def main(model_path, backend, **kwargs):
    start = time.perf_counter()
    if backend == 'turbomind':
        pipe = apply_turbomind(model_path, **kwargs)
    else:
        pipe = apply_pytorch(model_path, **kwargs)
    end = time.perf_counter()
    print(f'building pipeline cost: {end - start} s')
    
    input('press any key to continue')
    image = load_image('https://raw.githubusercontent.com/open-mmlab/mmdeploy/main/tests/data/tiger.jpeg')
    response = pipe(('describe this image', image))
    print(response)
    input('press any key to exit')
    
if __name__ == "__main__":
    fire.Fire(main)
    

```

File: lmdeploy/model.py [hunk 4]
```diff
@@ -1529,6 +1529,37 @@ def match(cls, model_path: str) -> Optional[str]:
             return 'glm4'
 
 
+@MODELS.register_module(name='internvl-phi3')
+class InternVLPhi3(Phi3Instruct):
```
Comment (lvhan028) [inline lmdeploy/model.py:1533]: May add UT for this class.

**Review by lvhan028 — APPROVED** [review]

**Comment by xin0623** [comment]
lmdeploy-5.0.0+cu118-cp39   Is this version supported? Why do I still get the same error?

**Comment by lvhan028** [comment]
This PR just was merged into the main branch but hasn't been released yet. 
We are going to release v0.5.1 around 15th, July.

**Comment by v3ss0n** [comment]
I had build using this PR , and when inference i got this error :

Mini-InternVL-Chat-4B-V1-5

AssertionError: turbomind does not support /home/nozander/.cache/huggingface/hub/models--OpenGVLab--Mini-InternVL-Chat-4B-V1-5/snapshots/920f7428f246dbb460a1d90c4a3ee9dc696158cc. Plz try pytorch engine instead.

**Comment by RunningLeon** [comment]
@v3ss0n hi, pls. use `lmdeploy check_env` to check your env info and post a sample code to reproduce.

**Comment by v3ss0n** [comment]
```
import numpy as np
import torch
import torchvision.transforms as T
from decord import VideoReader, cpu
from PIL import Image
from torchvision.transforms.functional import InterpolationMode
from transformers import AutoModel, AutoTokenizer

IMAGENET_MEAN = (0.485, 0.456, 0.406)
IMAGENET_STD = (0.229, 0.224, 0.225)


def build_transform(input_size):
    MEAN, STD = IMAGENET_MEAN, IMAGENET_STD
    transform = T.Compose([
        T.Lambda(lambda img: img.convert('RGB') if img.mode != 'RGB' else img),
        T.Resize((input_size, input_size), interpolation=InterpolationMode.BICUBIC),
        T.ToTensor(),
        T.Normalize(mean=MEAN, std=STD)
    ])
    return transform


def find_closest_aspect_ratio(aspect_ratio, target_ratios, width, height, image_size):
    best_ratio_diff = float('inf')
    best_ratio = (1, 1)
    area = width * height
    for ratio in target_ratios:
        target_aspect_ratio = ratio[0] / ratio[1]
        ratio_diff = abs(aspect_ratio - target_aspect_ratio)
        if ratio_diff < best_ratio_diff:
            best_ratio_diff = ratio_diff
            best_ratio = ratio
        elif ratio_diff == best_ratio_diff:
            if area > 0.5 * image_size * image_size * ratio[0] * ratio[1]:
                best_ratio = ratio
    return best_ratio


def dynamic_preprocess(image, min_num=1, max_num=6, image_size=448, use_thumbnail=False):
    orig_width, orig_height = image.size
    aspect_ratio = orig_width / orig_height

    # calculate the existing image aspect ratio
    target_ratios = set(
        (i, j) for n in range(min_num, max_num + 1) for i in range(1, n + 1) for j in range(1, n + 1) if
        i * j <= max_num and i * j >= min_num)
    target_ratios = sorted(target_ratios, key=lambda x: x[0] * x[1])

    # find the closest aspect ratio to the target
    target_aspect_ratio = find_closest_aspect_ratio(
        aspect_ratio, target_ratios, orig_width, orig_height, image_size)

    # calculate the target width and height
    target_width = image_size * target_aspect_ratio[0]
    target_height = image_size * target_aspect_ratio[1]
    blocks = target_aspect_ratio[0] * target_aspect_ratio[1]

    # resize the image
    resized_img = image.resize((target_width, target_height))
    processed_images = []
    for i in range(blocks):
        box = (
            (i % (target_width // image_size)) * image_size,
            (i // (target_width // image_size)) * image_size,
            ((i % (target_width // image_size)) + 1) * image_size,
            ((i // (target_width // image_size)) + 1) * image_size
        )
        # split the image
        split_img = resized_img.crop(box)
        processed_images.append(split_img)
    assert len(processed_images) == blocks
    if use_thumbnail and len(processed_images) != 1:
        thumbnail_img = image.resize((image_size, image_size))
        processed_images.append(thumbnail_img)
    return processed_images


def load_image(image_file, input_size=448, max_num=6):
    image = Image.open(image_file).convert('RGB')
    transform = build_transform(input_size=input_size)
    images = dynamic_preprocess(image, image_size=input_size, use_thumbnail=True, max_num=max_num)
    pixel_values = [transform(image) for image in images]
    pixel_values = torch.stack(pixel_values)
    return pixel_values


path = 'OpenGVLab/Mini-InternVL-Chat-2B-V1-5'
model = AutoModel.from_pretrained(
    path,
    torch_dtype=torch.bfloat16,
    low_cpu_mem_usage=True,
    trust_remote_code=True).eval().cuda()

tokenizer = AutoTokenizer.from_pretrained(path, trust_remote_code=True)
# set the max number of tiles in `max_num`
pixel_values = load_image('./examples/image1.jpg', max_num=6).to(torch.bfloat16).cuda()

generation_config = dict(
    num_beams=1,
    max_new_tokens=1024,
    do_sample=False,
)

# single-image single-round conversation (单图单轮对话)
question = """<image>
How many metalframes are there? Calculate carefully"""
response = model.chat(tokenizer, pixel_values, question, generation_config)
print(f'User: {question}')
print(f'Assistant: {response}')

```

**Comment by zhyncs** [comment]
ref https://github.com/zhyncs/lmdeploy-build/releases/tag/9f3e748

**Comment by v3ss0n** [comment]
ok gonna try that build

**Comment by v3ss0n** [comment]
```
LLVM ERROR: Cannot select: intrinsic %llvm.nvvm.shfl.sync.bfly.i32
zsh: IOT instruction (core dumped)  lmdeploy serve api_server OpenGVLab/Mini-InternVL-Chat-4B-V1-5
```

**Comment by v3ss0n** [comment]
I tried calling APi via api_server , and then i got coredumped.

**Pull request merged**